### PR TITLE
http set-node-ready implementation

### PR DESCRIPTION
**Pull request by lindlind**

### CODE REVIEW

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/impl/kuber/KuberNodeController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            .maxAttempts(3)
+            .intervalFunction(IntervalFunction.ofExponentialBackoff(1000))
+            .retryOnException(e -> e instanceof KubernetesClientException ke && !KuberUtils.isNotRetryable(ke))
+            .build();
+        retry = Retry.of("k8s-client-retry", retryConfig);
+    }
+
+    @Override
+    public void addLabels(String clusterId, String nodeName, Map<String, String> labels) {
+        final ClusterRegistry.ClusterDescription cluster;
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/impl/kuber/KuberNodeController.java:35]: let's add logs

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        String nodeName = requestBody.get("node_name");
+
+        if (clusterId == null || nodeName == null) {
+            String errorMessage = "Required data is not specified";
+            handleError(clusterId, nodeName, errorMessage);
+            return HttpResponse.status(HttpStatus.BAD_REQUEST);
+        }
+
+        try {
+            nodeController.addLabels(clusterId, nodeName, Map.of("lzy.ai/node-ready-to-use", "true"));
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:39]: let's declare all k8s labels in the `KuberLabels` class

**Review by imakunin — CHANGES_REQUESTED** [review]

File: lzy/allocator/src/main/java/ai/lzy/allocator/alloc/impl/kuber/KuberNodeController.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        if (node == null) {
+            throw new IllegalArgumentException("Node not found");
+        }
+
+        return node;
+    }
+
+    @Override
+    public void addLabels(String clusterId, String nodeName, Map<String, String> labels) {
+        LOG.debug("Adding labels; clusterId: {}, nodeName: {}, labels: {}", clusterId, nodeName, labels);
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/alloc/impl/kuber/KuberNodeController.java:59]: `labels` is a Map, but `Map::toString()` is not human friendly :)

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    @Inject
+    NodeController nodeController;
+
+    @Inject
+    AllocatorMetrics metrics;
+
+    @Inject
+    HttpClientAddressResolver clientAddressResolver;
+
+    Set<String> readyNodes = new HashSet<>();
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:45]: private final

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    HttpClient httpClient = HttpClient.newBuilder()
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:47]: private final

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            return HttpResponse.status(HttpStatus.BAD_REQUEST);
+        }
+
+        LOG.info("Set node ready request; clientAddress: {}, clusterId: {}, nodeName: {}",
+            clientAddress, clusterId, nodeName);
+
+        if (readyNodes.contains(nodeName)) {
+            String errorMessage = "Node was set ready before";
+            handleError(clientAddress, clusterId, nodeName, errorMessage);
+            return HttpResponse.status(HttpStatus.BAD_REQUEST);
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:71]: I suppose `HttpStatus.CONFLICT` is more suitable here

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            .toList();
+        try {
+            return nodeAddresses.contains(InetAddress.getByName(address));
+        } catch (UnknownHostException e) {
+            throw new IllegalArgumentException("Failed to parse client address");
+        }
+    }
+
+    private boolean ensureNodeIsReady(String host) {
+        final var requestUrl = "http://" + HostAndPort.fromParts(host, NODE_DAEMONSET_PORT) + "/isReady";
```
Comment (imakunin) [inline lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:121]: let's move it to the constant pattern/template `private static final String CHECK_NODE_READINESS_TEMPLATE = "http://%s/isReady"`

File: lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java (hunk 6 above)
Comment (lindlind) [reply lzy/allocator/src/main/java/ai/lzy/allocator/services/KuberNodeService.java:72]: I don't think we should reply different statuses on errors. It isn't necessary for daemonset, and is definitely not needed for another unexpected client.

**Review by imakunin — APPROVED** [review]

**Pull request merged**